extensions/agroal/runtime/src/main/java/io/quarkus/agroal/runtime/DataSources.java

```diff
@@ -274,8 +281,15 @@ private void applyNewConfiguration(AgroalDataSourceConfigurationSupplier dataSou
             TransactionIntegration txIntegration = new NarayanaTransactionIntegration(transactionManager,
                     transactionSynchronizationRegistry, null, false,
                     dataSourceJdbcBuildTimeConfig.transactions == io.quarkus.agroal.runtime.TransactionIntegration.XA
-                            ? xaResourceRecoveryRegistry
-                            : null);
+                            && transactionRuntimeConfig.enableXaRecovery
+                                    ? xaResourceRecoveryRegistry
+                                    : null);
+            if (dataSourceJdbcBuildTimeConfig.transactions == io.quarkus.agroal.runtime.TransactionIntegration.XA
+                    && !transactionRuntimeConfig.enableXaRecovery) {
+                log.warnv(
+                        "datasource {0} enables XA but transaction recovery is not enabled. Please use quarkus.transaction-manager.enable-xa-recovery=true otherwise you will have a risk to lose the data consistency when the application crash happens.",
```

Comment:
```suggestion
                        "Datasource {0} enables XA but transaction recovery is not enabled. Please use `quarkus.transaction-manager.enable-xa-recovery=true` otherwise data may be lost if the application is abruptly terminated",
```